pytorch_lightning/trainer/configuration_validator.py

```diff
@@ -69,6 +69,25 @@ def __verify_train_loop_configuration(self, model):
                 ' `training_step()`, `train_dataloader()` and `configure_optimizers()` to be defined.'
             )
 
+        has_optimizer_step = is_overridden('optimizer_step', model)
+        enable_pl_optimizer = self.trainer._enable_pl_optimizer
+        automatic_optimization = self.trainer.train_loop.automatic_optimization
+        if has_optimizer_step and not enable_pl_optimizer and automatic_optimization:
+            log.warn(
+                "When overriding `LightningModule` optimizer_step with"
+                " `Trainer(..., enable_pl_optimizer=False, automatic_optimization=True, ...)`,"
+                " we won t be calling `.zero_grad` we can't assume when you call your `optimizer.step()`."
+                " For Lightning to take care of it, please use `Trainer(enable_pl_optimizer=True)`."
+            )
+
+        going_to_accumulate_grad_batches = self.trainer.accumulation_scheduler.going_to_accumulate_grad_batches()
+
+        if has_optimizer_step and going_to_accumulate_grad_batches and automatic_optimization:
+            raise MisconfigurationException(
+                'When overriding `LightningModule` optimizer_step with `Trainer(automatic_optimization=True, ...)`,'
+                ' `accumulate_grad_batches` should to be 1. It ensures `optimizer_step` is called on every batch'
```

Comment:
Yes, it will. 
However, when people are overriding optimizer_step. They might not pay attention to batch_idx indices.
They might do something like this. 
```
Trainer(accumulate_grad_batches=2)
...
def optimizer_step(...):
     batch_idx % 3 ==0: optimizer.step()
```
which will result in accumulated gradient of 6 and not 2 as expected.
It is why I found safer to make sure the `optimizer_step` is called on every batch_idx.